src/Data/Fuzzy/Fuzzy.purs

```diff
@@ -0,0 +1,222 @@
+module Data.Fuzzy where
+
+import Prelude
+
+import Data.Array (snoc, unsnoc)
+import Data.Either (Either(..))
+import Data.Foldable (all, foldl)
+import Data.Generic.Rep (class Generic)
+import Data.Generic.Rep.Eq (genericEq)
+import Data.Generic.Rep.Ord (genericCompare)
+import Data.Generic.Rep.Show (genericShow)
+import Data.StrMap (StrMap, values)
+import Data.Maybe (Maybe(..), fromMaybe)
+import Data.Monoid (mempty)
+import Data.Newtype (class Newtype, unwrap)
+import Data.String (Pattern(..), drop, indexOf, lastIndexOf, take, toLower)
+import Data.String.Utils (length, toCharArray, words)
+import Data.Tuple (Tuple(..))
+
+newtype Fuzzy a = Fuzzy
+  { original :: a
+  , result   :: StrMap (Maybe Result)
+  , score    :: Rank
+  }
+
+derive instance genericFuzzy :: Generic (Fuzzy a) _
+derive instance newtypeFuzzy :: Newtype (Fuzzy a) _
+instance eqFuzzy :: Eq a => Eq (Fuzzy a) where eq = genericEq
+instance showFuzzy :: Show a => Show (Fuzzy a) where show = genericShow
+instance ordFuzzy :: Eq a => Ord (Fuzzy a) where
+  compare (Fuzzy { score }) (Fuzzy { score: score' }) = compare score score'
+
+newtype FuzzyStr = FuzzyStr
+  { result :: Result
+  , score  :: Rank
+  }
+
+derive instance genericFuzzyStr :: Generic FuzzyStr _
+derive instance newtypeFuzzyStr :: Newtype FuzzyStr _
+instance eqFuzzyStr :: Eq FuzzyStr where eq = genericEq
+instance showFuzzyStr :: Show FuzzyStr where show = genericShow
+instance ordFuzzyStr :: Ord FuzzyStr where
+  compare (FuzzyStr { score }) (FuzzyStr { score: score' }) = compare score score'
+
+data Depth = Full | Word | Char
+
+derive instance genericDepth :: Generic Depth _
+instance eqDepth :: Eq Depth where eq = genericEq
+instance showDepth :: Show Depth where show = genericShow
+
+data Pos = Start | Prefix | Mid | Suffix | End
+
+derive instance genericPos :: Generic Pos _
+instance eqPos :: Eq Pos where eq = genericEq
+instance showPos :: Show Pos where show = genericShow
+
+data Rank = Rank Int Int Int Int Int Int | None
+
+derive instance genericRank :: Generic Rank _
+instance eqRank :: Eq Rank where eq = genericEq
+instance showRank :: Show Rank where show = genericShow
+instance ordRank :: Ord Rank where compare = genericCompare
+
+instance semiringRank :: Semiring Rank where
```

Comment:
This isn't a law-abiding Semiring instance (see https://pursuit.purescript.org/packages/purescript-prelude/3.0.0/docs/Data.Semiring#t:Semiring).

For instance, 

**Annihilation:**  `zero * a = a * zero = zero`

yet: `mul (Rank 0 1 0 1 0 1) zero = None`

Almost all the laws are fine (associativity, right- and left- distributivity over multiplication, etc.) but the zero provides the problem. Is it possible to give this a law-abiding Semiring instance (and thus opt-in to the guarantees of the typeclass)?

Reply:
If `zero == None` then doesn't that abide by the annihilation law?

Reply:
It does feel a little strange to me, but I can’t come up with any counterexamples ... but perhaps a `Monoid` instance would be sufficient ? `scoreDistance` could be refactored to not use multiplication (`s/1/d/`), and it looks like the rest of it uses addition anyways, which would indicate to me that the semantics of it fit well with an (additive) monoid. Just because it’s numeric doesn’t mean multiplication makes sense ;) (e.g. 2D or 3D Vectors can be combined additively, but not really multiplied ... you can scale them with a group action though!)

Reply:
Hey @MonoidMusician thanks for the feedback! I used [purescript-quickcheck-laws](https://github.com/garyb/purescript-quickcheck-laws) at your suggestion and it does seem* to abide by the `Semiring` laws, but I rather like your idea to use `Monoid` instance instead, and might just go with that

\* I say it seems to abide by the laws bc I'm not sure it manages full coverage. My instance for `Arbitrary` is defined as following:

```
instance arbitraryRank :: Arbitrary Rank
  where
    arbitrary = Rank
      <$> arbitrary
      <*> arbitrary
      <*> arbitrary
      <*> arbitrary
      <*> arbitrary
      <*> arbitrary
```
But that will never generate `None` (`Rank` is defined as `newtype Rank = Rank Int Int Int Int Int Int | None`) so I'm not certain my coverage is complete (although `zero = None` and assumedly the `Semiring` tests involve `zero`)

Reply:
sure, I think i can see how that makes sense ... btw https://pursuit.purescript.org/packages/purescript-quickcheck/4.6.1/docs/Test.QuickCheck.Gen#v:frequency if you want ;)